Add setup relayer node guide - Closes 1534 (#1750)
LiskHQ/lisk-docs: docs/modules/ROOT/pages/run-blockchain/setup-relayer.adoc

= Setting up a relayer node
Muhammad Talha <muhammad.talha@lightcurve.io>
:description: How to setup a relayer node.
:toc: preamble
:idprefix:
:idseparator: -
:sectnums:
:docs_sdk: v6@lisk-sdk::

// // Project URLS
:url_guides_reg_sidechain: build-blockchain/register-sidechain.adoc#how-to-register-a-sidechain-on-the-mainchain
:url_guides_reg_mainchain: build-blockchain/register-sidechain.adoc#how-to-register-the-mainchain-on-the-sidechain

====
On this page, you'll learn how to:

* [x] How to set up a relayer node on a sidechain?
* [x] How to set up a relayer node on a mainchain?
====

The process of setting up a relayer node on a mainchain and sidechain are almost identical.
The subsequent sections describe the steps to set up a relayer node.

.Prerequisites
[NOTE]
====
To use this guide, it is assumed that the following criteria have been met:

. A sidechain has been registered on the mainchain, as explained in the guide: xref:{url_guides_reg_sidechain}[How to register a sidechain on the mainchain?]

> . The mainchain has been registered on the sidechain, as described in the guide: xref:{url_guides_reg_mainchain}[How to register the mainchain on the sidechain?] 

Author: Since registration is the last step for a node to be connected to the blockchain, I think the pre-requisite is accurate as it tells exactly what needs to be done for a node to be connected to the blockchain, and since a node operator or a validator will most probably be registering the chains I think it suffices to say that registration is necessary. 
If blockchain registers, it will automatically mean that the node of the blockchain are connected to the network and API access to nodes will be possible on receiving chain.

> . The mainchain has been registered on the sidechain, as described in the guide: xref:{url_guides_reg_mainchain}[How to register the mainchain on the sidechain?] 

Author: Please share the links that you think are suitable here:
> Managing a node connected to the mainchain (link to Lisk Core)
Managing a node connected to the sidechain

> . The mainchain has been registered on the sidechain, as described in the guide: xref:{url_guides_reg_mainchain}[How to register the mainchain on the sidechain?] 

Reviewer: 1. I would link to this section: run-blockchain/index.html#how-to-set-up-a-lisk-mainnet-node
2. I am not sure we have a fitting guide for this. Also, the setup might be different on different sidechains, so maybe we can just add a footnote to refer to the setup guides of the respective sidechain.

> . The mainchain has been registered on the sidechain, as described in the guide: xref:{url_guides_reg_mainchain}[How to register the mainchain on the sidechain?] 

Reviewer: Sorry, I do not see the requirement to have a running node in the text.
Also, I don't see why the sidechain relayer has prerequisite to habe the status `active` and `registered`, but not the mainchain relayer. Imho there is no need to separate between the two. For both, it is necessary to complete the registration process as explained in the guide (which already includes both, sc and mc registration)

+
[tabs]
=====
Sidechain::
+
--
[source,bash]
----
sidechain_client> lisk.cryptography.encrypt.stringifyEncryptedMessage(await lisk.cryptography.encrypt.encryptMessageWithPassword('0d7501d3d5c9accaefb3c0b6a569473b59391ae406f6324f98fa6dd70e119368a6454f898d3b82c41b158206c72ecfe917a1071c8084b496a0c5867afc10830b', 'lisk'))

'kdf=argon2id&cipher=aes-256-gcm&version=1&ciphertext=57db80457db93a1abeceee5c6f951ca04579c447a06f45cf5e8b5398e207a26da53a6b191a02c479ede455950eacb48f32d6609f2cd4b5a1a51e895b210b587ef046e6c3151ef2212efd0808b45328742d09a279e7d667f1670ff02a2fd5c91f4afd0a08efb8e6e90b0b11e93b15da8daaeea543a0ff54f3dd51c66cac3b04c6&mac=7822258b12e0c787f5bd622c562914438a9d74ca1e11e11b840f3001a678b04f&salt=d4d051a123326ad2b82c022603e790b6&iv=0bb9e76cd5163d6c5af9d89d&tag=fbcdb355b5135d48df948841de5fcdf5&iterations=1&parallelism=4&memorySize=2024'
----
--
+
Mainchain::
+
--
[source,bash]
----
lisk-core> lisk.cryptography.encrypt.stringifyEncryptedMessage(await lisk.cryptography.encrypt.encryptMessageWithPassword('0d7501d3d5c9accaefb3c0b6a569473b59391ae406f6324f98fa6dd70e119368a6454f898d3b82c41b158206c72ecfe917a1071c8084b496a0c5867afc10830b', 'lisk'))

> 'kdf=argon2id&cipher=aes-256-gcm&version=1&ciphertext=57db80457db93a1abeceee5c6f951ca04579c447a06f45cf5e8b5398e207a26da53a6b191a02c479ede455950eacb48f32d6609f2cd4b5a1a51e895b210b587ef046e6c3151ef2212efd0808b45328742d09a279e7d667f1670ff02a2fd5c91f4afd0a08efb8e6e90b0b11e93b15da8daaeea543a0ff54f3dd51c66cac3b04c6&mac=7822258b12e0c787f5bd622c562914438a9d74ca1e11e11b840f3001a678b04f&salt=d4d051a123326ad2b82c022603e790b6&iv=0bb9e76cd5163d6c5af9d89d&tag=fbcdb355b5135d48df948841de5fcdf5&iterations=1&parallelism=4&memorySize=2024'

Reviewer: Please separate command and response in the code snippets for better overview

}
----
--
=====
Once the configurations are ready, update the *config.json* file of the respective client such as mainchain or sidechain.

=== Register plugin
Once the *config.json* has been updated, it is required to register the plugin with the client.
The process differs for both sidechain and mainchain.

[tabs]
=====
Sidechain::
+
--
On a sidechain, the *app.ts* file of the client can be updated to have the following options:

.sidechain_client/src/app/app.ts
[source,typescript]
----
import { Application, PartialApplicationConfig } from 'lisk-sdk';
// Import the 'ChainConnectorPlugin'
import { ChainConnectorPlugin } from '@liskhq/lisk-framework-chain-connector-plugin';
import { registerModules } from './modules';
import { registerPlugins } from './plugins';

export const getApplication = (config: PartialApplicationConfig): Application => {
	const { app } = Application.defaultApplication(config);

	registerModules(app);
	registerPlugins(app);
    // Register the ChainConnectorPlugin with the app
	app.registerPlugin(new ChainConnectorPlugin());
	return app;
};
----

The client must be rebuilt to entertain the changes to the code.

[source,bash]
----
npm run build
----

> Alternatively, the plugin can also be enabled using the `--enable-chain-connector-plugin` flag whilst starting the sidechain client, without updating the *app.ts* file.

Reviewer: Is this the alternative or the recommended way to enable the plugin?

Author: It is an alternative way, there is no recommended way TBH. A user can go with any given method to enable the plugin as they like. Both are correct ways to do the same thing.

Reviewer: I see. Maybe we can just change order of the two options, as this one looks more simple, but without recommending one option directly?

Author: I have made the adjustment as suggested.

+
Mainchain::
+
--
.Creating an encrypted key via mainchain client
[source,bash]
----
lisk-core> lisk.cryptography.encrypt.stringifyEncryptedMessage(await lisk.cryptography.encrypt.encryptMessageWithPassword('0d7501d3d5c9accaefb3c0b6a569473b59391ae406f6324f98fa6dd70e119368a6454f898d3b82c41b158206c72ecfe917a1071c8084b496a0c5867afc10830b', 'lisk'))
----

.Encrypted key
[source,bash]
----
'kdf=argon2id&cipher=aes-256-gcm&version=1&ciphertext=57db80457db93a1abeceee5c6f951ca04579c447a06f45cf5e8b5398e207a26da53a6b191a02c479ede455950eacb48f32d6609f2cd4b5a1a51e895b210b587ef046e6c3151ef2212efd0808b45328742d09a279e7d667f1670ff02a2fd5c91f4afd0a08efb8e6e90b0b11e93b15da8daaeea543a0ff54f3dd51c66cac3b04c6&mac=7822258b12e0c787f5bd622c562914438a9d74ca1e11e11b840f3001a678b04f&salt=d4d051a123326ad2b82c022603e790b6&iv=0bb9e76cd5163d6c5af9d89d&tag=fbcdb355b5135d48df948841de5fcdf5&iterations=1&parallelism=4&memorySize=2024'
----
--
=====

. The `encryptMessageWithPassword` function will return an encrypted key, which should be added to the config of the blockchain.

=== Prepare config for chain connector plugin
There's a slight difference in the `chainConnector` config of a sidechain and a mainchain. 

> The mainchain expects the `chainID` of the receiving chain along with the mandatory fields.

Reviewer: Can you please point me to where it is mentioned exactly? And why does the guide say `chainID`, while the Readme says `receivingChainID`?

> The mainchain expects the `chainID` of the receiving chain along with the mandatory fields.

Author: In this section, and yes it should be `receivingChainID` 👍🏼 
https://lightcurve.atlassian.net/wiki/spaces/SDK/pages/6095082/Chain+Connector+Plugin+Usage?NO_SSR=1#Prepare-config-for-chain-connector-plugin

    "password": "lisk", // The password used to create the encryptedPrivateKey
    "ccuFee": "100000000",
    "receivingChainIPCPath": "~/.lisk/relayer",
    "receivingChainID": "04000002"
}
----
--
=====
Once the configuration is ready, update the *config.json* file of the respective client such as mainchain or sidechain.

=== Register plugin
Once the *config.json* has been updated, it is required to register the plugin with the client.
The process differs for both sidechain and mainchain.

[tabs]
=====
Sidechain::
+
--
On a sidechain, the plugin can be enabled using the `--enable-chain-connector-plugin` flag whilst starting the sidechain client.

[source,bash]
----
./bin/run --enable-chain-connector-plugin --overwrite-config
----

Alternatively, the *app.ts* file of the client can be updated to have the following options:

> .sidechain_client/src/app/app.ts

Author: This plugin is registered in the `app.ts` as per the following guide:
https://lightcurve.atlassian.net/wiki/spaces/SDK/pages/6095082/Chain+Connector+Plugin+Usage?NO_SSR=1

> .sidechain_client/src/app/app.ts

Reviewer: I think the guide is not  following best practise in this case :)

};
----

The client must be rebuilt to entertain the changes to the code.

[source,bash]
----
npm run build
----
The client can then be started with the following command:

[source,bash]
----
./bin/run --overwrite-config
----

--
+
Mainchain::
+
--
On a mainchain, the plugin can be enabled using the `--enable-chain-connector-plugin` flag whilst starting the mainchain client.

[source,bash]
----
lisk-core --enable-chain-connector-plugin --overwrite-config
----
--
=====
Since the config of the chain is updated during the process, the node operator must update the existing config with the `--overwrite-config` flag.

=== Result
Once the client is running, the node operators should see the following log messages, depending on the type of client.

[tabs]
=====
Sidechain::
+
--
[source,bash]
----
2023-03-17T14:42:30.426Z INFO XYZ.local application 96733 No valid CCU can be generated for the height: 58
----
--
+
Mainchain::
+
--
[source,bash]
----
2023-03-17T14:42:30.426Z INFO XYZ.local plugin_chainConnector 96899 No valid CCU can be generated for the height: 58
----
--
=====
Since we just set up a relayer node and haven't sent a CCU/CCM, the aforementioned log messages are expected.

> The messages suggest that the blockchain doesn't have any finalized block height for which we can create a certificate and there are no pending CCUs/CCMs to send across the chain.

Reviewer: Isn't a first CCM created during the registration process on both chains? So that it can be expected that there is already 1 CCM present directly after registration? And if a node operator enables the chain connector plugin after the registration, how likely is it, that there are no CCMs to process when the plugin is enabled the first time? In case the default scenario is, that there are CCMs to be processed, so that it doesn't log the described error, it should explain the success message instead.

Author: AFAIK there is no CCM after registration. @ishantiw maybe can help us clarify this?

Reviewer: I am pretty sure there is: https://liskhq.github.io/lisk-docs/beta/understand-blockchain/interoperability/sidechain-registration-and-recovery.html#registration-ccm

Author: Is there an endpoint or method which can be used to get details of such CCM?

Reviewer: You can check the outbox size of the channel data store: https://liskhq.github.io/lisk-docs/beta/build-blockchain/register-sidechain.html#check-the-sidechain-channel 

Not sure if you can look at the individual CCMs, though 🤔

Author: Hmmm! I do get the size of the outbox: 1, not sure what it means for us though.

**Sidechain**
```
./bin/run endpoint:invoke interoperability_getChannel '{"chainID": "04000000"}' --pretty                                                                              
{
  "messageFeeTokenID": "0400000000000000",
  "outbox": {
    "appendPath": [
      "b947f78350104a72381044ae1a239752084215e0d5e2fc1cb0eba8956ec4761a"
    ],
    "root": "b947f78350104a72381044ae1a239752084215e0d5e2fc1cb0eba8956ec4761a",
    "size": 1
  },
  "inbox": {
    "appendPath": [],
    "root": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
    "size": 0
  },
  "partnerChainOutboxRoot": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
  "minReturnFeePerByte": "1000"
}
```
**Mainchain**
```
lisk-core endpoint:invoke interoperability_getChannel '{"chainID": "04000002"}' --pretty                                                                               
{
  "messageFeeTokenID": "0400000000000000",
  "outbox": {
    "appendPath": [
      "1a8fd0bea33f8144853e4f86ea57c38e84046dd7def2776e6d6782b90357eae8"
    ],
    "root": "1a8fd0bea33f8144853e4f86ea57c38e84046dd7def2776e6d6782b90357eae8",
    "size": 1
  },
  "inbox": {
    "appendPath": [],
    "root": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
    "size": 0
  },
  "partnerChainOutboxRoot": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
  "minReturnFeePerByte": "1000"
}
```

There is no success message for me, I get the error message. ⬇️
> In case the default scenario is, that there are CCMs to be processed, so that it doesn't log the described error, it should explain the success message instead.

Reviewer 2: When we register a chain, a registration CCM is created and added to the outbox but if there is no valid certificate to be created, then we wait for for the finalized height, meanwhile log "No valid CCU can be generated for the height: <>". Once the first finalized height is created then we can include the CCM with the first CCU and send it across, for subsequent CCUs, we can have a scenario where there are pending CCMs but no certificate, in that case we can still send a CCU with only pending CCMs and without certificate.

Reviewer: Just to clarify, so in case the sidechain is newly registered, and we set up the first relayer, then this log message is expected, right? But for a running sidechain, a new relayer would be already able to send CCUs because the finalized height is reached? I think we need to find and explain a way to verify that the relayer was set up correctly for every use case. This guide is not only for people setting up a relayer after registration, but more how to setup a relayer in general.

Author: @Tschakki Good point. I agree. @ishantiw Is there a way to realize what Mona suggests?

Also if you guys agree, I think we should have a separate guide about sending CCU/CCMs and should also describe any edge cases a node operator has to face. This guide can be part two of the set-up relayer guide.

Reviewer: > separate  guide about sending CCU/CCMs

I don't think there are many manual steps included after activating the plugin - what exactly do you imagine to be explained in the guide?

Author: Not sure, as I am not familiar with the process. Discussing various ways of sending CCU/CCM (if there are any) can be explained there. Edge cases can be addressed as well. We can also show various responses like for a successful CCU/CCM, for failure, etc. More clarity can be achieved once we work on the topic.

Reviewer: Yes I agree, let's keep that in mind and ocme back later to the topic, once we have more practical knowledge :)

            "password": "lisk",
            "ccuFee": "100000000",
            "receivingChainIPCPath": "~/.lisk/lisk-core"
        }
    }
}
----
--
=====
IMPORTANT: A node operator must add either the value of `*receivingChainIPCPath*` or `*receivingChainWsURL*` in the chain connector's config.

== Enabling the Chain Connector plugin
A node operator can perform the following steps to enable the chain connector plugin and turn a node into a relayer node.

=== Creating an encrypted private key
. The first step is to create an encrypted private key. 
A node operator can use a REPL session to call Lisk's cryptography libraries.
+
[tabs]
=====
Sidechain::
+
--
[source,bash]
----
./bin/run console
Entering Lisk REPL: type `Ctrl+C` or `.exit` to exit
----
--
+
Mainchain::
+
--
[source,bash]
----
lisk-core console
Entering Lisk REPL: type `Ctrl+C` or `.exit` to exit
----
--
=====

> . Call the `encrypt` function to create an encrypted private key and pass the `generatorPrivateKey` of your account to it.

Reviewer 2: ```suggestion
. Call the `encrypt` function to create an encrypted private key and password the `generatorPrivateKey` of your account to it.
```

> . Call the `encrypt` function to create an encrypted private key and pass the `generatorPrivateKey` of your account to it.

Reviewer 2: Ah this is not generatorPrivateKey but ed private key, you can simply say to encrypt private key of any account on the other chain that you want to use as a relayer address and has balance to sign and send the transaction

+
Mainchain::
+
--
.Creating an encrypted key via mainchain client
[source,bash]
----
lisk-core> lisk.cryptography.encrypt.stringifyEncryptedMessage(await lisk.cryptography.encrypt.encryptMessageWithPassword('0d7501d3d5c9accaefb3c0b6a569473b59391ae406f6324f98fa6dd70e119368a6454f898d3b82c41b158206c72ecfe917a1071c8084b496a0c5867afc10830b', 'lisk'))
----

.Encrypted key
[source,bash]
----
'kdf=argon2id&cipher=aes-256-gcm&version=1&ciphertext=57db80457db93a1abeceee5c6f951ca04579c447a06f45cf5e8b5398e207a26da53a6b191a02c479ede455950eacb48f32d6609f2cd4b5a1a51e895b210b587ef046e6c3151ef2212efd0808b45328742d09a279e7d667f1670ff02a2fd5c91f4afd0a08efb8e6e90b0b11e93b15da8daaeea543a0ff54f3dd51c66cac3b04c6&mac=7822258b12e0c787f5bd622c562914438a9d74ca1e11e11b840f3001a678b04f&salt=d4d051a123326ad2b82c022603e790b6&iv=0bb9e76cd5163d6c5af9d89d&tag=fbcdb355b5135d48df948841de5fcdf5&iterations=1&parallelism=4&memorySize=2024'
----
--
=====

. The `encryptMessageWithPassword` function will return an encrypted key, which should be added to the config of the blockchain.

=== Prepare config for chain connector plugin
There's a slight difference in the `chainConnector` config of a sidechain and a mainchain. 

> The mainchain expects the `chainID` of the receiving chain along with the mandatory fields.

Reviewer 2: on both mainchain and sidechain you need chainID of the receiving chain

    "password": "lisk", // The password used to create the encryptedPrivateKey
    "ccuFee": "100000000",
    "receivingChainIPCPath": "~/.lisk/relayer",
    "receivingChainID": "04000002"
}
----
--
=====
Once the configuration is ready, update the *config.json* file of the respective client such as mainchain or sidechain.

=== Register plugin
Once the *config.json* has been updated, it is required to register the plugin with the client.
The process differs for both sidechain and mainchain.

[tabs]
=====
Sidechain::
+
--
On a sidechain, the plugin can be enabled using the `--enable-chain-connector-plugin` flag whilst starting the sidechain client.

[source,bash]
----

> ./bin/run --enable-chain-connector-plugin --overwrite-config

Reviewer 2: ```suggestion
./bin/run start --enable-chain-connector-plugin --overwrite-config
```

----

Alternatively, the *plugins.ts* file of the client can be updated to have the following options:

.sidechain_client/src/app/plugins.ts
[source,typescript]
----
import { Application } from 'lisk-sdk';
// Import the 'ChainConnectorPlugin'
import { ChainConnectorPlugin } from '@liskhq/lisk-framework-chain-connector-plugin';

export const registerPlugins = (app: Application): void => {
    // Register the ChainConnectorPlugin with the app
    app.registerPlugin(new ChainConnectorPlugin());
};

----

The client must be rebuilt to entertain the changes to the code.

[source,bash]
----
npm run build
----
The client can then be started with the following command:

[source,bash]
----

> ./bin/run --overwrite-config

Reviewer 2: ```suggestion
./bin/run start --overwrite-config
```

= Setting up a relayer node
Muhammad Talha <muhammad.talha@lightcurve.io>
:description: How to setup a relayer node.
:toc: preamble
:idprefix:
:idseparator: -
:sectnums:
:docs_sdk: v6@lisk-sdk::

// // Project URLS
:url_guides_reg_sidechain: build-blockchain/register-sidechain.adoc
:url_guides_reg_mainchain: build-blockchain/register-sidechain.adoc#how-to-register-the-mainchain-on-the-sidechain
:url_relayer_node: run-blockchain/index.adoc#relayer-nodes
:url_mainnet_node: run-blockchain/index.adoc#how-to-set-up-a-lisk-mainnet-node

:fn_mainnet: footnote:witness[For more information about maintaining a Lisk Mainnet node, see the xref:{url_mainnet_node}[How to set up a Lisk Mainnet node] section.]

====
On this page, you'll learn:

* [x] How to set up a relayer node on a sidechain?
* [x] How to set up a relayer node on a mainchain?
====

The process of setting up a xref:{url_relayer_node}[relayer node] on a mainchain and sidechain are almost identical.
The subsequent sections describe the steps to set up a relayer node.

> .Prerequisites
> [NOTE]
> ====
> To use this guide, it is assumed that the following criteria have been met:
> 
> .For a sidechain to set up a relayer node:
> * The sidechain must be registered to the mainchain as described in the xref:{url_guides_reg_sidechain}[] guide.
> * The sidechain account on the mainchain must have either the `registered` or `active` status.
> 
> .For a mainchain to set up a relayer node:
> * The mainchain should be registered on the sidechain to which it intends to communicate as described in the xref:{url_guides_reg_mainchain}[How to register the mainchain on the sidechain] guide{fn_mainnet}.

Author: Does this sound okay to you?
```suggestion
.Prerequisites
[NOTE]
====
To use this guide, it is assumed that the following criteria have been met:

* The sidechain must be registered to the mainchain as described in the xref:{url_guides_reg_sidechain}[] guide.
* The sidechain account on the mainchain{fn_mainnet} must have either the `registered` or `active` status.
```

Reviewer: It is going in the right direction, but as mentioned before, I would put in the prerequisites that they need to have nodes on sidechain and mainchain.

1. Managing a node connected to the mainchain (link to Lisk Core)
2. Managing a node connected to the sidechain
3. The sidechain should be registered to the mainchain according to the guide "How to register a sidechain"(link) and the status of the sidechain account should be `registered` or `active`.

Author: Done.

> 

Reviewer: Sorry, I do not see the requirement to have a running node in the text.
Also, I don't see why the sidechain relayer has prerequisite to have the status `active` and `registered`, but not the mainchain relayer. Imho there is no need to separate between the two. For both, it is necessary to complete the registration process as explained in the guide (which already includes both, sc and mc registration)

Author: > Sorry, I do not see the requirement to have a running node in the text.

Without a running node, it will not be possible to even do the registration. The reader is already aware or should be aware of running a node, before setting up the relayer node. IMHO this is a pre-req for the registration guide.

> Also, I don't see why the sidechain relayer has a prerequisite to have the status active and registered, but not the mainchain relayer. Imho there is no need to separate the two. 

Because the liveness condition is only for a sidechain and not for a mainchain. A mainchain is going to stay alive.
https://github.com/LiskHQ/lips/blob/main/proposals/lip-0045.md#life-cycle-of-a-sidechain

The separation of prerequisites is intentional as either a sidechain node operator will read it or a mainchain node operator, to set up the relayer. That's why content is separated for mainchain and sidechain, throughout the page.

Author: I have made a suggestion above, please consider that.

====

// // A sidechain has been registered on the mainchain, as explained in the guide: xref:{url_guides_reg_sidechain}[How to register a sidechain on the mainchain?]
// . The mainchain has been registered on the sidechain, as described in the guide: xref:{url_guides_reg_mainchain}[How to register the mainchain on the sidechain?] 
== Installing the Chain Connector plugin

The chain connector plugin comes pre-installed with any chain which has been initialized with Lisk SDK v6.
However, in case it is required to install it manually, run the following command in the root folder of the client.

[source,bash]
----
npm install @liskhq/lisk-framework-chain-connector-plugin
----

> == Configuring the Chain Connector plugin

Reviewer: This is more like an explanation, and there are no concrete steps that a user can do here. I would therefore merge this content with the section where we are actually configuring the plugin, under `=== Prepare config for chain connector plugin`

Author: I tried that and this breaks the flow for me as in the suggested case, the introduction of the concept `encryptedPrivateKey` will come after the instruction: **Creating an encrypted private key**. If you agree, I would like to keep it.

> With these configurations, the details such as the receiver’s chainID, the receiving chain’s path, CCU frequency, encrypted private key, etc. can be set.

Reviewer 2: ```suggestion
With these configurations, the details such as the receiving chain’s chainID and it's IPC path or WS URL, CCU frequency, encrypted private key, etc. can be set.
```

[tabs]
=====
Configuration options::
+
--
[cols="2,2,2,2",options="header",stripes="hover"]
|===
|Name
|Type
|Description
|Sample

|`receivingChainIPCPath`
|string
|The IPC path of a receiving node
|~/.lisk/lisk-core

|`receivingChainWsURL`
|string
|The WS URL of a receiving node
|ws://142.93.230.246:4002/rpc-ws

|`ccuFrequency`
|integer
|Number of blocks after which a CCU should be created
|1

|`encryptedPrivateKey`
|string
|Encrypted privateKey of the relayer
|kdf=argon2id&cipher=aes-256-gcm&version=1& ciphertext=da36ef64cbf2687d62b0-
14dafdfa8ef8c823b2b1562ae78819599080e4500529-
b75e80093fba066879f0767e0de83abe285efb259dd9be5109b8a4ef66cfc52ec613314586c-
1aa1da3a6737c0f8b7f0de7fb4d1b85860cd23915bbcee774e1d85b357e342816a917e517f7-
c702e1a1deb28dd69a4b69ae2ac67a5c4c4236101c&-
mac=e253decce05dd50758400d5c7408532a162fedf583ff9cafcb7ad3e12f6b8011&-
salt=40ab2dcdd387e4372ed1dbb948a9ef84&iv=4559dcaee67eb2c1a0957ecf&-
tag=81c2c332d915454bed4be26018c598c5&iterations=1&parallelism=4&-
memorySize=2024

|`ccuFee`
|string
|The fee to be paid for each CCU transaction
|100000000

|`password`
|string
|Password to decrypt encryptedPrivateKey
|lisk

> |`isSaveCCU`

Reviewer 2: I have recently added ccuSaveLimit which by default is 300 CCUs and user can set it to any value to tell how many CCUs it wants to save https://github.com/LiskHQ/lisk-sdk/pull/8315

[source,bash]
----
npm run build
----
The client can then be started with the following command:

[source,bash]
----
./bin/run start --overwrite-config
----

--
+
Mainchain::
+
--
On a mainchain, the plugin can be enabled using the `--enable-chain-connector-plugin` flag whilst starting the mainchain client.

[source,bash]
----
lisk-core start --network alphanet --enable-chain-connector-plugin --overwrite-config   
----
--
=====
Since the config of the chain is updated during the process, the node operator must update the existing config with the `--overwrite-config` flag.

=== Result
Once the client is running, the node operators should see the following log messages, depending on the type of client.

[tabs]
=====
Sidechain::
+
--
[source,bash]
----
2023-03-17T14:42:30.426Z INFO XYZ.local application 96733 No valid CCU can be generated for the height: 58
----
--
+
Mainchain::
+
--
[source,bash]
----
2023-03-17T14:42:30.426Z INFO XYZ.local plugin_chainConnector 96899 No valid CCU can be generated for the height: 58
----
--
=====
Since we just set up a relayer node and haven't sent a CCU/CCM, the aforementioned log messages are expected.

> The messages suggest that the blockchain doesn't have any finalized block height for which we can create a certificate and there are no pending CCUs/CCMs to send across the chain.

Reviewer 2: ```suggestion
The messages suggest that the blockchain doesn't have any finalized block height for which we can create a certificate or there are no pending CCUs/CCMs to send across the chain.
```
